### PR TITLE
Fix cuda Transpose bug 16039

### PR DESCRIPTION
**Pull request by RyanUnderhill**
### Description
Transpose will fail in cuda for FLOAT16 for tensors larger than 1048x1048 due to our optimized case exceeding the cuda grid size of 65536.

The fix is to just use our regular cuda transpose in these cases.

### Motivation and Context
https://github.com/microsoft/onnxruntime/issues/16039

### CODE REVIEW

File: onnxruntime/core/providers/cuda/tensor/transpose.cc [hunk 1]
```diff
@@ -196,7 +196,7 @@ Status Transpose::DoTranspose(const cudaDeviceProp& prop,
     auto mn = TryTransposeWithCublas(new_permutations, new_input_dims);
     int M = std::get<0>(mn);
     int N = std::get<1>(mn);
-    if (M != 0 && N != 0) {
+    if (M != 0 && N != 0 && (element_type != ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT16 || CanUse_cublasTransposeHelper_MLFloat16(M))) {
```
Comment (hariharans29) [inline onnxruntime/core/providers/cuda/tensor/transpose.cc:199]: Need to pass both M and N here ?

File: onnxruntime/core/providers/cuda/fpgeneric.cu [hunk 2]
```diff
@@ -64,11 +64,22 @@ __global__ void CopyVectorBFloat16(const onnxruntime::BFloat16* x, int incx, onn
...
 }  // namespace
 
+dim3 cublasTransposeHelperDimGrid(int m, int n) {
+  return {(n + TRANS_TILE_DIM - 1) / TRANS_TILE_DIM, (m + TRANS_TILE_DIM - 1) / TRANS_TILE_DIM, 1};
+}
+
+// cublasTransposeHelper can only be used if it won't overflow the 65536 grid dimension size
+__host__ bool CanUse_cublasTransposeHelper_MLFloat16(int m, int n) {
+  dim3 dimGrid = cublasTransposeHelperDimGrid(m, n);
+  return dimGrid.x < 65536 && dimGrid.y < 65536;
```
Comment (hariharans29) [inline onnxruntime/core/providers/cuda/fpgeneric.cu:74]: Is there a way to programmatically get the max blocks per grid device property and check against that ? Like device.maxBlocksPerGrid ? Or is it always bound to be 65535 ?

File: onnxruntime/core/providers/cuda/tensor/transpose.cc (hunk 1 above)
Comment (RyanUnderhill) [reply onnxruntime/core/providers/cuda/tensor/transpose.cc]: Yep! Hadn't built yet.

File: onnxruntime/core/providers/cuda/fpgeneric.cu (hunk 2 above)
Comment (RyanUnderhill) [reply onnxruntime/core/providers/cuda/fpgeneric.cu:74]: There likely is, the table here shows that it varies for x, but not for y. x goes up to 2^31-1 starting with compute 3.0. y is 65535 max for all versions, and z is 3 for all versions:

https://en.wikipedia.org/wiki/CUDA#Version_features_and_specifications

You could probably modify it to take advantage of that, but it's not a few hour fix.

**Comment by tianleiwu** [comment]
Add a test case for cuda and rcom?

**Comment by RyanUnderhill** [comment]
> 

Done, I verified it hits the cuda fix in the code.

File: onnxruntime/core/providers/cuda/fpgeneric.cu (hunk 2 above)
Comment (tianleiwu) [reply onnxruntime/core/providers/cuda/fpgeneric.cu:74]: The condition can be `dimGrid.y < 65536`. The x has larger limit. Did you test a case of dimGrid.x > 65536 && dimGrid.y < 65536?

https://en.wikipedia.org/wiki/Thread_block_(CUDA_programming)#:~:text=Blocks%20can%20be%20organized%20into,y%20and%20z%20dimensions%20respectively.

Comment (RyanUnderhill) [reply onnxruntime/core/providers/cuda/fpgeneric.cu]: Fixed. I tried adding a test case for the dimGrid.x but couldn't decipher the shape/perm requirements to make it happen. It kept making dimGrid.x very low with what I tried.

**Review by hariharans29 — APPROVED** [review]

**Review by tianleiwu — APPROVED** [review]

**Pull request merged**